packages/typespec-vscode/src/vscode-cmd/openapi3-preview.ts

```diff
@@ -0,0 +1,306 @@
+import { readdir, rm } from "fs/promises";
+import * as semver from "semver";
+import * as vscode from "vscode";
+import logger from "../log/logger.js";
+import { getBaseFileName, getDirectoryPath, joinPaths } from "../path-utils.js";
+import { TspLanguageClient } from "../tsp-language-client.js";
+import { createTempDir, parseOpenApi3File, throttle } from "../utils.js";
+
+const TITLE = "Preview in OpenAPI3";
+
+export async function showOpenApi3(
+  uri: vscode.Uri,
+  context: vscode.ExtensionContext,
+  client: TspLanguageClient,
+) {
+  const selectedFile = uri?.fsPath ?? vscode.window.activeTextEditor?.document.uri.fsPath;
+  if (!selectedFile || !selectedFile.endsWith(".tsp")) {
+    logger.error("Please select a Typespec file", [], {
+      showOutput: true,
+      showPopup: true,
+    });
+    return;
+  }
+
+  const mainTspFile = selectedFile.endsWith("main.tsp") ? selectedFile : await getMainTspFile();
+  if (mainTspFile === undefined) {
+    logger.error(`No 'main.tsp' file can be determined from '${selectedFile}' in workspace.`, [], {
+      showOutput: true,
+      showPopup: true,
+    });
+    return;
+  }
+
+  await loadOpenApi3PreviewPanel(mainTspFile, context, client);
+}
+
+async function getMainTspFile(): Promise<string | undefined> {
+  const files = await vscode.workspace.findFiles("**/main.tsp").then((uris) => {
+    return uris.filter((uri) => uri.scheme === "file" && !uri.fsPath.includes("node_modules"));
+  });
+
+  switch (files.length) {
+    case 0:
+      logger.error("No 'main.tsp' file found in the workspace.", [], {
+        showOutput: true,
+        showPopup: true,
+      });
+      return undefined;
+    case 1:
+      return files[0].fsPath;
+    default:
+      return await vscode.window.showQuickPick(
+        files.map((file) => file.fsPath),
+        {
+          title: TITLE,
+          placeHolder: "Select the 'main.tsp' file",
+        },
+      );
+  }
+}
+
+const openApi3TempFolders = new Map<string, string>();
+const openApi3PreviewPanels = new Map<string, vscode.WebviewPanel>();
+const selectedOpenApi3OutputFiles = new Map<string, string>();
+
+async function loadOpenApi3PreviewPanel(
+  mainTspFile: string,
+  context: vscode.ExtensionContext,
+  client: TspLanguageClient,
+) {
+  const compilerVersion = client.initializeResult?.serverInfo?.version;
+  if (!compilerVersion || semver.lt(compilerVersion, "0.65.0")) {
+    logger.error("OpenAPI3 preview requires TypeSpec compiler version 0.65.0 or later.", [], {
+      showOutput: true,
+      showPopup: true,
+    });
+    return;
+  }
+
+  if (openApi3PreviewPanels.has(mainTspFile)) {
+    // if panel is already opened, there should be a watcher to automatically update the content
+    // no need to generate the openapi3 files
+    const panel = openApi3PreviewPanels.get(mainTspFile)!;
+    const outputFolder = await getOutputFolder(mainTspFile);
+    if (!outputFolder) {
+      panel.dispose();
```

Comment:
also delete the panel from map? otherwise, the disposed panel will be used next time, right?

Reply:
In the `onDidDispose()` event handler, the panel will be removed from the map.
```
 panel.onDidDispose(() => {
      openApi3PreviewPanels.delete(mainTspFile);
      selectedOpenApi3OutputFiles.delete(mainTspFile);
      watch.dispose();
    });
```

I debug test it, `dispose()` can trigger `onDidDispose()` event handler.